src/System.Private.ServiceModel/tests/Scenarios/Client/ChannelLayer/DuplexChannelShapeTests.4.0.0.cs

```diff
@@ -140,30 +140,22 @@ public static void CallbackBehavior_ConcurrencyMode_Single_NetTcpBinding()
         InstanceContext instanceContext;
         DuplexChannelFactory<IWcfDuplexService_CallbackConcurrencyMode> factory = null;
         IWcfDuplexService_CallbackConcurrencyMode channel = null;
-        try
-        {
-            // *** SETUP *** \\
-            binding = new NetTcpBinding(SecurityMode.None);
-            instanceContext = new InstanceContext(new CallbackHandler_ConcurrencyMode_Single());
-            factory = new DuplexChannelFactory<IWcfDuplexService_CallbackConcurrencyMode>(instanceContext, binding, Endpoints.DuplexCallbackConcurrencyMode_Address);
 
-            // *** EXECUTE *** \\
-            channel = factory.CreateChannel();
-            channel.DoWork();
+        // *** SETUP *** \\
```

Comment:
I don't see how removing the try block fixed this. I believe you just got lucky when you submitted it again. What's happening is the manual reset event is reset at the end of the method call which then returns to the service. That then makes a second call back to the client and the client is being closed at the same time as another call is coming in. There's a race condition and you happened to avoid it when the tests got ran again. What you need to do is save the task from channel.DoWork() (which needs renaming to DoWorkAsync as it's an async method) and await it before moving to cleanup. That way you you know both callbacks have completed and there will be no error.

Reply:
I was trying to code according to my understanding of your comments, but it doesn't seem to work as expected, here is the code snippet:

```
        channel = factory.CreateChannel();
        Task task = channel.DoWorkAsync();

        Assert.True(imp.MyManualResetEvent.WaitOne(20000));
        Assert.Equal(1, imp.Counter);

        await task;
```
the await statement didn't wait for completion of the two tasks defined in DoWorkAsync which is awaiting Task.WhenAll(t1,t2).

Reply:
The call to DoWorkAsync kicks off two async callbacks on the service side. In the service code you save each of the tasks and then await Task.WhenAll which shouldn't complete until both of the callback tasks have completed. One the await Task.WhenAll completes, the service side DoWorkAsync completes and sends the result (which just says it finished without an exception) to the client. On the client side the await on channel.DoWorkAsync task will complete when the server returns that completed message to the client. At that point there should be no requests in progress initiated by either the server or the client. The `await task` statement should wait for completion of the two tasks defined in DoWorkAsync.  
I'll pull down your changes and see if I can work out what's going on.

Reply:
At server side, the line below doesn't await but return immediately:
` await Task.WhenAll(t1, t2);`